docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,485 @@
+
+If you don't provide a select operator, all fields parsed will be returned by default. This means all fields from a log line if you're using implicit extraction. However if you used explicit extraction, only those fields will be returned.
+
+For example this query:
+
+```logql
+{job="prod/query-frontend"} | logfmt | sort duration desc | limit 25 | select query duration
+```
+
+is the same as this one
+
+```logql
+{job="prod/query-frontend"} | logfmt query duration | sort duration desc | limit 25
+```
+
+because the field extraction was explicit, only those fields (`query` and `duration`) are outputted.
+
+This is why select is important to use for large queries. Select will allows users to remove data (column) in their logs that are useless, repetitive and not interesting for the current investigation.
+
+The select operation will also support the [go templating language](https://golang.org/pkg/text/template/) when supplied with string argument and not fields.
+
+Fields extracted will be available under the `.FieldName` and `$FieldName` notation in the template.
+
+This way users can format each field to be more digestible. Returning a payload size  field with value of `12321421230123` is less meaningful than `12TB`.
+
+Example:
+
+```logql
+{job="prod/query-frontend"} | logfmt | select payload_size="{{ .size | humanize }}" level="{{ $level | ToUpper}}"
+```
+
+Renaming is allowed with or without templates, and is optional. An autogenerated field name will be used if it can't be inferred. (`field1`,`field2`, etc...)
+
+__In this current stance, select is not greedy meaning if you want to rename or template a single field, you'll have to declare all other fields to keep them.__
+
+In the future we might consider a `exclude` operator to take all fields but those passed as parameter. (`{job="prod/query-frontend"} | logfmt | exclude msg`)
+
+### Limit
+
+The limit operator (`| limit x`) is a new operator to overwrite the query string limit, this can be handy for building dashboard with different limits but also when writing and testing queries, you could start to experiment with a low limit to increase the feedback loop.
+
+If the limit operator is missing in the expression, the limit from the query string is used.
+
+The limit operator can be applied to a log field  as well as a log stream.
+
+For example all of the above are correct.
+
+```logql
+{job="prod/query-frontend"} | logfmt query duration | sort duration desc | limit 25
+{job="prod/query-frontend"} | limit 10
+{job="prod/query-frontend"} |= "foo" | limit 20
+```
+
+Using limit operator with metric queries will result into an error. (e.g `rate({job="prod/query-frontend"} | limit 10[5m])`)
+
+### Uniq
+
+The uniq operator allows to filter out duplicates. Without any parameter (`| uniq`) each fields combination must be unique from one entry to another. However
+the filtering can be forced on a specific field, for example `| uniq status_code`, this will returns one entry per status code.
```

Comment:
How do you decide which entry to keep if the matching is done only on a subset of the log entry (ie. a single field)?

Reply:
It means it will keep an entry only if the field has not been seen before. This would allows you to get the list of unique address IP as explained by @slim-bean . `{app="nginx"} | glob ... | uniq address_ip`

I did a poor job at explaining all those uses cases. Gonna do another run to explain that for each operators.